sdk-setup/src/mb2

```diff
@@ -964,6 +964,61 @@ make_gvariant_array() {
     printf "[%s]\n" "$body"
 }
 
+global_lock_init() {
+    [[ $MB2_SELF_TEST_RUN ]] && return
+
+    # Lock files need to be available from sb2, hence the use of $HOME.
+
+    # Two lock files are used in order to allow child processes to acquire the
+    # lock on behalf of mb2. This is done by interchanging the main lock file.
+    #
+    # Special care needs to be taken when the two mechanisms need to be nested.
+    # See how EXTENDED_PRE_INSTALL_LOCK_ARGS is overriden locally after calling
+    # global_lock_acquire at some places.
+    GLOBAL_LOCK=$HOME/.$ME.lock
+    GLOBAL_LOCK_INTERCHANGE_LOCK=$HOME/.$ME.lock.lock
+
+    # A child process may interchange the actual lock for this one, pre-acquired
+    # by mb2 and released only when mb2 exits.
+    #
+    # TODO global cleanup
+    trap '[[ $GLOBAL_LOCK_EXTENDED_LOCK ]] && rm -f "$GLOBAL_LOCK_EXTENDED_LOCK"' EXIT
+    GLOBAL_LOCK_EXTENDED_LOCK=$(mktemp "$HOME/.$ME.lock.XXX") || return
+    exec {GLOBAL_LOCK_EXTENDED_LOCK_FD}>"$GLOBAL_LOCK_EXTENDED_LOCK"
+    flock "$GLOBAL_LOCK_EXTENDED_LOCK_FD"
+
+    # Keep in sync with `rpm --showrc`
+    local def_spec_install_pre_head='%{___build_pre}'
+    local def_spec_install_pre_tail='[ "$RPM_BUILD_ROOT" != "/" ] && rm -rf "${RPM_BUILD_ROOT}"\
+mkdir -p `dirname "$RPM_BUILD_ROOT"`\
+mkdir "$RPM_BUILD_ROOT"\
+%{nil}'
+
+    # Let rpmbuild acquire the lock at the beginning of the %install section
+    # (before the shared build root or a shared output directory is accessed)
+    EXTENDED_PRE_INSTALL_LOCK_ARGS=(--define "__spec_install_pre \
+$def_spec_install_pre_head\\
+flock ${GLOBAL_LOCK_INTERCHANGE_LOCK@Q} flock ${GLOBAL_LOCK@Q} \\\\\\
+    ln -f ${GLOBAL_LOCK_EXTENDED_LOCK@Q} ${GLOBAL_LOCK@Q}\\
+$def_spec_install_pre_tail")
+}
```

Comment:
This is a bit difficult to follow, but I guess that can't be really helped; A diagram would help a lot, but of course that's a bit difficult to include in a comment. But please verify that I understood this correctly: The aim is that that once the build process reaches the install step, the already locked extended lock will replace the global lock, and it will only be freed after mb2 exits. Meaning that if we have multiple build processes running concurrently, they will happily continue, possibly locking and freeing the global lock as they go, but once one of the processes reaches the install step, the other processes will wait at the next global_lock_acquire() or beginning of install step until the first process has finished completely. I.e. After the lock has been interchanged, global_lock will remain locked until the mb2 process which locked it has been finished.

Reply:
Yup, this ensures that no other build will mess up the RPM build root between %install and the actual package creation, which is done by various processes spawned by rpmbuild and by rpmbuild itself. So the lock is owned by sfdk but only comes into the play with the %install section.